### ISSUE TITLE
Corrupted Audio in Key Word Detection

### ISSUE THREAD

**Issue by EliasLum** [alexa/avs-device-sdk#1804]
**Overview** - We can see corrupted audio when logging the audio stream to a .pcm file during the KWD.
***Device* - Raspberry Pi 3
**OS and build** - Raspbian GNU/Linux 9 (stretch)
**Alexa-SDK version** - 1.18.0

## Summary
In order to check the input to the wake word engine we logged the 10ms buffer of the wake word engine to a .pcm file. While doing so we could observe that the audio is corrupted. It can be seen (and heard) that at the beginning/end of the 160 sample buffer there seems to be unexpected values that result in crackling and noises that corrupt the audio file.

## What is the expected behavior?
We would expect a clean audio file.

## What behavior are you observing?
A corrupted audio file with unexplicable corruption at the beginning/end of the 10ms-buffers (160 samples). See attached .pcm-file.

## Steps to reproduce the issue:
We could reproduce the behavior for the standard SampleApp build. Only file changed SensoryKeywordDetector.cpp to add the logging capability.
in the line before -  while (!m_isShuttingDown):
`FILE *fIn = fopen("/home/pi/sdk-folder/LOGS/Sensory_InputLog_16khz.pcm", "w+b");`
in the line before -  snsrClearRC(m_session):
`fwrite(audioDataToPush.data(), sizeof(int16_t), audioDataToPush.size(), fIn); `
in the line before - m_streamReader->close():
` fclose(fIn);`

## List of Investigations
Following a list of already done investigations is given:
- removed all detection functionality -> only print out the .pcm file in the detection loop. The file is still corrupted.
- Checked the pcm-dump when running AudioInputProcessorIntegrationTest.cpp, where a local file is used as input instead of microphone input. The pcm-dump was uncorrupted.
- Ran AudioInputProcessorIntegrationTest.cpp (logs attached)
- Ran a short Sensory command "what's the time" with DEBUG9 (log attached)
- Checked the Alexa voice history: https://www.amazon.com/alexaprivacysettings. The files there are not corrupted which indicates that generally recording works (at least the unblocking stream, blocking stream might have a problem).
- recordings with arecord are not corrupted

## Attachements
[GithubTicket.zip](https://github.com/alexa/avs-device-sdk/files/5341618/GithubTicket.zip)

**Comment by mvelegon-amzn** [comment]
Hi @EliasLum would you be able to log audio to a file inside AudioInputProcessor as well? I couldn't find anything different in the stream we send to the KWD and the Cloud so I would like to see if you are noticing a difference.

**Comment by EliasLum** [comment]
Hey @mvelegon-amzn, 
First of all thanks for the reply. Yes of course i can log an audio file inside the AudioInputProcessor. Could you maybe give me some indications where i could read the stream from to log the audio?
I tried to add the logging to: avs-device-sdk\CapabilityAgents\AIP\src\AudioInputProcessor.cpp
but i couldn't find out how to get the data vector of the audio. I guess i would have to somehow read provider.stream or m_reader?
e.g. could you maybe give me an indication where and how to get a data-vector i can log to an audio-file in AudioInputProcessor.cpp?

**Comment by EliasLum** [comment]
@mvelegon-amzn,
Do you have any thoughts/indications considering my previous comment?

**Comment by EliasLum** [comment]
Dear Alexa Team,

I am a bit surprised by your inactivaty on this issue. However, in order to make sure the problem is not connected to my implementation i set up the raspberry and SDK from scratch. So I did not do, install or change anything apart from the below mentioned steps:

1. Erased the SD card and imaged the newest Raspbian Buster OS using: https://www.raspberrypi.org/blog/raspberry-pi-imager-imaging-utility/ 
2. Started up the rPi
3. Went to the console and installed cmake: sudo apt-get install cmake
4. Followed all the instructions on: https://developer.amazon.com/en-US/docs/alexa/avs-device-sdk/raspberry-pi.html
e.g. ran the lines in the terminal (excluding Bluetooth functionality)
5. Added the three lines from my first comment to log a .pcm file in: SensoryKeywordDetector.cpp
6. Started up the app and spoke some test sentences

When i look at the .pcm file generated it still sounds corrupted (as described in the first attachment). I did not do anything apart from the above mentioned steps so I am puzzled where the issue lies.

Are you able to reproduce the described behaviour?
If we want to take a look at the audio inside AudioInputProcessor , can you give me support on my previous questions?

**Comment by kclchan** [comment]
Hi @EliasLum, you said you added your code before `while (!m_isShuttingDown)` and tried to write audioDataToPush into a file, please note that audioDataToPush is just an empty vector and nothing has been push to it yet.  Maybe you'd like to try again by adding you lines of code after `readFromStream` is called, and use wordsRead for how many bytes you want to write to your file.  Thanks!

**Comment by EliasLum** [comment]
Hey @kclchan,
Please read my comments more closely. I open a file before `while (!m_isShuttingDown) `  however writing audioDataToPush into a file of course happens after reading from stream. As mentioned this happens before `snsrClearRC(m_session)`. How did you imagine we have got any audio at all when writing an empty vector to a file...?
I was assuming that the size of the stream would be constantly 160 samples (10ms at 16000kHz). Which is also the size of audioDataToPush. From what you mention i suspect it might be possible that less than 160 samples are written to audioDataToPush per iteration and we should only write the amount of bytes defined in wordsRead.

**Comment by kclchan** [comment]
I am glad that this is working for you now.  Closing the issue.
